File: python/kohaku/src/kohaku/spec/predicate.py
"""Pure functions that evaluate a visibleWhen predicate against state (port of TS predicate.ts).

Leaf comparisons are done by equality of canonical JSON strings — matched deterministically down to key
order and array/object structure (in takes this equality element by element).

When state has no initial value for the referenced key (equivalent to undefined), it never matches the
comparison value (including null). This distinguishes "unspecified undefined" from "value is null". Note
that unknown keys in a visibleWhen leaf ref are forbidden in advance by STATE_REF_UNKNOWN, so at actual
render time a key is never unspecified.
"""

from __future__ import annotations

from .canonical_json import UNDEFINED, _Undefined, canonical_stringify
from .models import (
    AllPredicate,
    AnyPredicate,
    JsonValue,
    LeafPredicate,
    NotPredicate,
    VisibleWhen,
)


def evaluate_visible_when(pred: VisibleWhen, state: dict[str, JsonValue]) -> bool:
    if isinstance(pred, AllPredicate):
        return all(evaluate_visible_when(p, state) for p in pred.all)
    if isinstance(pred, AnyPredicate):
        return any(evaluate_visible_when(p, state) for p in pred.any)
    if isinstance(pred, NotPredicate):
        return not evaluate_visible_when(pred.not_, state)
    return _evaluate_leaf(pred, state)


def _evaluate_leaf(pred: LeafPredicate, state: dict[str, JsonValue]) -> bool:
    key = pred.ref[len("$state.") :]
    current: JsonValue | _Undefined = state.get(key, UNDEFINED)
    if pred._is_set("eq"):
        return _same_json(current, pred.eq)
    if pred._is_set("ne"):
        return not _same_json(current, pred.ne)
    if pred._is_set("in_"):
        assert pred.in_ is not None
        return any(_same_json(current, v) for v in pred.in_)
    # Numeric comparison: always false when the state value is not a number (undefined / null / string, etc.).
    if pred._is_set("gt"):
        assert pred.gt is not None
        return _is_number(current) and current > pred.gt  # type: ignore[operator]
    if pred._is_set("lt"):
        assert pred.lt is not None
        return _is_number(current) and current < pred.lt  # type: ignore[operator]
    if pred._is_set("gte"):
        assert pred.gte is not None
        return _is_number(current) and current >= pred.gte  # type: ignore[operator]
    if pred._is_set("lte"):
        assert pred.lte is not None
        return _is_number(current) and current <= pred.lte  # type: ignore[operator]
    # Existence check: match "whether the state value is other than null/undefined" against the truth of exists.
    if pred._is_set("exists"):
        present = not isinstance(current, _Undefined) and current is not None
        return present == pred.exists
    # Unreachable due to exactly-one validation, but for exhaustiveness default to visible.
    return True


def collect_state_refs(pred: VisibleWhen) -> list[str]:
    """Enumerate the state keys referenced by the predicate tree in traversal order (including duplicates)."""
    if isinstance(pred, AllPredicate):
        return [key for p in pred.all for key in collect_state_refs(p)]
    if isinstance(pred, AnyPredicate):
        return [key for p in pred.any for key in collect_state_refs(p)]
    if isinstance(pred, NotPredicate):
        return collect_state_refs(pred.not_)
    return [pred.ref[len("$state.") :]]


def _same_json(a: JsonValue | _Undefined, b: JsonValue) -> bool:
    # When a is undefined (no initial value in state), it never matches b (always a defined JSON value).
    if isinstance(a, _Undefined):
        return False
    return canonical_stringify(a) == canonical_stringify(b)


def _is_number(v: object) -> bool:
    """Equivalent to JS typeof v === "number" (bool is not a number)."""
    return isinstance(v, (int, float)) and not isinstance(v, bool)

File: packages/spec-core/src/predicate.ts
import { canonicalStringify } from "./canonical-json.js";
import type { JsonValue } from "./schema/json.js";
import type { LeafPredicate, VisibleWhen } from "./schema/state.js";

/**
 * A pure, environment-neutral function that evaluates a visibleWhen predicate against state.
 * Leaf comparison is done by equality of canonicalStringify outputs — matching deterministically down
 * to key order and array/object structure (in takes this equality per element). Compound predicates
 * (all / any / not) fold recursively. The comparison logic is centralized here so React and non-React
 * renderers share identical semantics.
 *
 * When state has no initial value for the key (undefined), it never matches the comparison value
 * (including null): canonicalStringify(undefined) returns JS undefined, which is always non-equal to
 * strings like `"null"`. This distinguishes "unspecified undefined" from "value is null". Note that an
 * unknown key in a VisibleWhen leaf ref is forbidden up front by STATE_REF_UNKNOWN, so current is never
 * undefined at actual render time.
 */
export function evaluateVisibleWhen(pred: VisibleWhen, state: Readonly<Record<string, JsonValue>>): boolean {
  if ("all" in pred) return pred.all.every((p) => evaluateVisibleWhen(p, state));
  if ("any" in pred) return pred.any.some((p) => evaluateVisibleWhen(p, state));
  if ("not" in pred) return !evaluateVisibleWhen(pred.not, state);
  return evaluateLeaf(pred, state);
}

/** Evaluates a leaf predicate (ref + exactly one comparison). */
function evaluateLeaf(pred: LeafPredicate, state: Readonly<Record<string, JsonValue>>): boolean {
  const key = pred.ref.slice("$state.".length);
  const current = state[key];
  if (pred.eq !== undefined) return sameJson(current, pred.eq);
  if (pred.ne !== undefined) return !sameJson(current, pred.ne);
  if (pred.in !== undefined) return pred.in.some((v) => sameJson(current, v));
  // Numeric comparison: always false if the state value is not a number (undefined / null / string, etc.).
  if (pred.gt !== undefined) return typeof current === "number" && current > pred.gt;
  if (pred.lt !== undefined) return typeof current === "number" && current < pred.lt;
  if (pred.gte !== undefined) return typeof current === "number" && current >= pred.gte;
  if (pred.lte !== undefined) return typeof current === "number" && current <= pred.lte;
  // Existence check: matches whether the state value is other than null/undefined against exists' boolean.
  if (pred.exists !== undefined) return (current !== undefined && current !== null) === pred.exists;
  // Unreachable because of refine, but default to visible for type exhaustiveness.
  return true;
}

/**
 * Enumerates the state keys a predicate tree references, in traversal order (including duplicates).
 * A pure function that lets validate.ts's STATE_REF_UNKNOWN check (referenced keys must have an initial
 * value in spec.state) inspect every leaf even for compound predicates.
 */
export function collectStateRefs(pred: VisibleWhen): string[] {
  if ("all" in pred) return pred.all.flatMap(collectStateRefs);
  if ("any" in pred) return pred.any.flatMap(collectStateRefs);
  if ("not" in pred) return collectStateRefs(pred.not);
  return [pred.ref.slice("$state.".length)];
}

function sameJson(a: JsonValue | undefined, b: JsonValue): boolean {
  // When a is undefined, canonicalStringify returns JS undefined, which does not match the string
  // representation of b (always a defined JSON value). This is what guarantees the undefined/null
  // distinction.
  return canonicalStringify(a) === canonicalStringify(b);
}
